### ISSUE TITLE
Clear the long error texts from Navigator

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/e2e#122]
I don't like that the Navigator contains long texts when I instantiate some sort of errors. I'll move those to separate error classes or just remove them. I'll think about it. I already though about creating an environment test, which runs by the first usage. These tests could make the UnsupportedEnvironment errors obsolete, and those errors contain the long template texts I don't like...

**Comment by inf3rno** [comment]
I am pretty sure I'll move these to self-tests / environment-tests classes and run them by requiring the lib. I am not sure what page I will load with them though. Maybe something with data URI I guess. If that test succeeds, then we can continue, otherwise the lib should throw an UnsupportedEnvironment error by using openWindow.

**Comment by inf3rno** [comment]
In the Navigator I have code like this:

```js
        await TaskManager.micro();
        this.currentNavigation = this.followingNavigation;
        this.followingNavigation = createExposedPromise();
        if (this.window.document.readyState !== "complete")
            return this.navigationError(new UnsupportedEnvironment("The ready state of the window must not change by next tick after unload."));
        this.trigger(this.events.page.unload);

        await TaskManager.micro();
        this.trigger(this.events.navigation.request);
        if (this.unexpectedNavigation)
            return this.navigationError(new UnexpectedNavigation());
        this.unexpectedNavigation = true;

        await TaskManager.macro();
        if (this.window.document.readyState !== "loading")
            return this.navigationError(new UnsupportedEnvironment(`The ready state of the window must be "loading" instead of "${this.window.document.readyState}" 10-20 msecs after unload.`));
        this.trigger(this.events.navigation.response, this.window.document.URL);

        await TaskManager.micro();
```

Checking the readyState is assertion. It is evident that the code won't work, when these assertions are failing, so it does not have any additional value in those terms, but it fails earlier with a meaningful message (and class name), so we know that the problem is with the environment and not with the library. There can be other errors, by which the problem is with the library or with the test. These assertions are mixed with those errors, that's what I don't like, because it is enough to run a self-test once to check the environment and after that there is no need for that while running the code. So I think the goal is ok.

a.) self testing
```js
        await SelfTest.run(); // runs only once, cached after that

        await TaskManager.micro();
        this.currentNavigation = this.followingNavigation;
        this.followingNavigation = createExposedPromise();
        this.trigger(this.events.page.unload);

        await TaskManager.micro();
        this.trigger(this.events.navigation.request);
        if (this.unexpectedNavigation)
            return this.navigationError(new UnexpectedNavigation());
        this.unexpectedNavigation = true;

        await TaskManager.macro();
        this.trigger(this.events.navigation.response, this.window.document.URL);

        await TaskManager.micro();
```

b.) assertion

```js
        const it = new SelfTest(this.window);
        await TaskManager.micro();
        this.currentNavigation = this.followingNavigation;
        this.followingNavigation = createExposedPromise();
        it.shouldNotLoadYet();
        this.trigger(this.events.page.unload);

        await TaskManager.micro();
        this.trigger(this.events.navigation.request);
        if (this.unexpectedNavigation)
            return this.navigationError(new UnexpectedNavigation());
        this.unexpectedNavigation = true;
        it.shouldNotLoadYet();

        await TaskManager.macro();
        it.shouldBeLoading();
        this.trigger(this.events.navigation.response, this.window.document.URL);

        await TaskManager.micro();
```

**Comment by inf3rno** [comment]
The a.) runs only once and does not cover the entire code, just a few minimalistic examples. The b.) can cover the entire code, but if so, then why are we writing tests? Hmm actually tests are checking whether the library does what we expect, while assertions are checking whether the environment does what we expect. They overlap in some cases, but for example our tests do not check whether the window.addEventListener exists, because we take that granted...

**Comment by inf3rno** [comment]
I don't like my current solution. I mean it almost does the same as the Navigator would by loading a data uri. It is like a big example. I'd like to split it up into smaller chunks, which check the browser differences I descibed in the wiki: https://github.com/inf3rno/e2e/wiki/browser-support-and-browser-features

**Comment by inf3rno** [comment]
I created a relative long checking with data uri, but Chrome does not seem to load the data uris with location.assign. Interesting, because I don't have an error message either. It just does not do anything.

**Comment by inf3rno** [comment]
https://www.fxsitecompat.com/en-CA/docs/2017/data-url-navigations-on-top-level-window-will-now-be-blocked/
We are not talking about the top level window here, but I guess they added other constraints meanwhile. In theory I could load plain text or image without an issue.

**Comment by inf3rno** [comment]
https://www.bleepingcomputer.com/news/security/firefox-will-block-navigational-data-uris-as-part-of-an-anti-phishing-feature/
This tells the same that plain text won't be affected.

**Comment by inf3rno** [comment]
I tried `about:about` etc. But loading it appears to by sync and does not lead to an unload event. I think I'll give up and do the self-test by the first load maybe.

**Comment by inf3rno** [comment]
I need to check whether the different readyStates, the document and the URIs are ok. I might give up and revert to assertions instead.

**Comment by inf3rno** [comment]
Hmm another option to give a self-check URI.

**Comment by inf3rno** [comment]
I don't like the self-check URI configuration. This should run in the background and should not require any developer configuration. Since data URIs no longer work and I need an address to test all I can do is using the first address they give.

**Comment by inf3rno** [comment]
The current code looks like this:

```js
    static async run() {
        try {
            this.checkBasicObjectsAndMethods();
            const childWindow = this.checkOpeningBlankPage();
            const listenerRemovalCheck = this.checkEventListenerRemovalByLoading(childWindow);
            listenerRemovalCheck.start();
            const dataURI = this.createEmptyPageDataURI();
            await this.checkLoadingAddress(childWindow, dataURI);
            await this.checkClosing(childWindow);
            listenerRemovalCheck.finish();
            return Promise.resolve();
        } catch (error) {
            return Promise.reject(new UnsupportedEnvironment(error));
        }
    }
```

**Comment by inf3rno** [comment]
- checkBasicObjectsAndMethods - checks whether the objects and methods needed for the lib are present
 - checkOpeningBlankPage - checks whether we can open a blank page and access its content
 - checkEventListenerRemovalByLoading - checks whether event listeners are removed by page load
 - checkLoadingAddress - checks whether the address is loaded properly and we got a new document with loading ready state
 - checkClosing - checks whether we got an unload event by closing

**Comment by inf3rno** [comment]
1. - sync, can be done instantly
 2. - sync normally, can be done instantly
 3. - requires a page load, but possibly loading about:about or about:blank will reveal it
 4. - requires page loading
 5. - does not require page loading, can be done by about pages too I think, not that important though

So I think we can do a page opening, loading about pages and close it as a first test. Another option to give that window for the first load, but I would rather avoid that if possible. Yet another option to have some sort of factory/semaphore that will allow us to reuse windows.

**Comment by inf3rno** [comment]
loading about: pages appears to by sync and the outcome will be a new document, interesting...

**Comment by inf3rno** [comment]
what about loading javascript URIs? if data URIs don't work maybe those will

**Comment by inf3rno** [comment]
I realized meanwhile that js URIs are just for running js code with the href attribute. I think I am out of options, it is not possible to do this without configuration or without loading the first user given URI. It is really bad, because all I can do is guessing when I got a SecurityError, because since v70 (or maybe 69) I cannot access the error page.

**Comment by inf3rno** [comment]
I think I'll end up with assertions, but I will create a SelfTest class or something like that, and do the assertions in the methods of that. For example

```js
    async leaving() {
        // ...
        await TaskManager.micro();
        this.assertUnloading(this.window, this.address);
        // ...
        await TaskManager.macro();
        this.assertLoading(this.window);
        // ...
    }
```

It does not cover every scenario though, just where we expect a state, that is environment dependent and important to us. Hmm actually we could listen to the events to test the environment, so we don't have to add extra code to the navigator and our new class will be more or less separate from it.

**Comment by inf3rno** [comment]
I don't like assertions, because if I use them, then what should be the limit? If I test everything and trust nothing in the environment, then 2/3 of the code will be try-catch and assertion. The only option to use the events as checkpoints and throw error if something does not work properly. This is a limited approach though. I have to check what tests I can do this way.

**Comment by inf3rno** [comment]
We need to check the content of the `checkLoadingAddress` method. If we can solve everything with events, then we can go that way. Otherwise it is better to forget on the fly environment checking and assume that the dev knows what they are doing.

**Comment by inf3rno** [comment]
```js
static checkLoadingAddress(childWindow, address) {
        const oldDocument = childWindow.document;
        const loadingPromise = createExposedPromise();
        const timer = setTimeout(function () {
            loadingPromise.reject("Loading data URI timed out. No unload event or sync page loading by data URI.");
        }, 500);
        try {
            childWindow.location.assign(address);
            childWindow.addEventListener("unload", async function unloaded() {
                childWindow.removeEventListener("unload", unloaded);
                clearTimeout(timer);
                await TaskManager.macro();
                try {
                    EnvironmentChecking.checkNewDocument(childWindow, oldDocument);
                    await EnvironmentChecking.checkReadyStateChange(childWindow);
                    EnvironmentChecking.checkStopping(childWindow);
                }
                catch (error) {
                    return loadingPromise.reject(error);
                }
                loadingPromise.resolve();
            });
        }
        catch (error) {
            clearTimeout(timer);
            loadingPromise.reject("Could not load data URI: " + error);
        }
        return loadingPromise;
    }
```

**Comment by inf3rno** [comment]
What do we have here?
 - listen to unload -> timeout
 - wait macro task
 - check whether we can access the new document, it is really new and its ready state is loading
 - wait for ready state change -> timeout
 - expect an interactive ready state to be the next
 - check window.stop (throws error when the doc is not accessible)

All of these can be done with the current events. Since I added many events and all of them are separated by macro or micro tasks we can easily do this I think.

**Comment by inf3rno** [comment]
I made up my mind about this, we need to test whether the self test does not throw an error, so we need a selftest test after all. This is ok, it should run without error so it is nothing complicated. I want to implement this to detect errors caused by Chrome changes.

**Comment by inf3rno** [comment]
https://bugs.chromium.org/p/chromium/issues/detail?id=902879
added an issue, maybe they make chrome error pages accessible again
I have to wait for a response and decide after that what happens to the current environment checking

**Comment by inf3rno** [comment]
possible solutions:
 1. no longer check the environment
 2. check the environment on the fly
    1. the dev should configure the env loading test with an address
    2. the first given address is used for the env loading test
 3. check the environment in the tests (use google.com for the env loading test)

possible implementations for solution 2x.
 1. solve the communication between the navigator and the env check for the loading test
     1. use the navigator events as checkpoints
     2. use assertions in the navigator
     3. call the env test directly from the navigator
 2. calling the env test by navigator instantiation and by the first navigation
     1. we keep the sync navigator creation, inject the checker and call it in an async hook by first load
     1. we keep the sync navigator creation and use a navigator proxy to run the checker by first load
     1. we add async navigator creation and run half of the tests by creation and half by first load

**Comment by inf3rno** [comment]
solutions:
The 1. is not acceptable:
We need to handle Chrome changes immediately if possible, so we need a test we can use to check the environment and identify the changes.

The 2.i. is not acceptable:
I don't want another unnecessary param.

The 2.ii. is not acceptable:
We got the SecurityError by Chrome error pages, so by a wrong first address we don't know whether the problem is with the environment or with the address. I don't think address problems will be rare compared to environment problems. So this is not something we can ignore.

The 3. is not acceptable:
We need something that tells the developer that the problem is with the Chrome version and not with the library. So these tests are not just for me, because I can easily identify the problem. I just run the tests that worked and if they fail, then it is pretty sure that Chrome changed again. But if the devs got some unrelated error, they will hardly identify the cause, and they will spend hours debugging the issue.

**Comment by inf3rno** [comment]
I think we could have a single engine that does the environment dependent things and has some sort of interface the high level navigator can use. This way it can wrap the environment dependent things and put them into try-catch blocks.

**Comment by inf3rno** [comment]
Another option to promisify the engine and do the following by load:

```js
await engine.open();
engine.validate(address);
engine.location.assign(address);
engine.started();
await engine.response();
// replace console with proxy?
await engine.interactive();
return engine.document;
```

**Comment by inf3rno** [comment]
I can live with this, but in theory, this is what the navigator should contain. It seems like we have another layer in the middle between the "engine" and the high level API. I am not too worried about this. The engine could contain the very minimum in the navigation topic and the navigator could have a richer interface like .submit(form) etc...

If so, then the barely minimum is not the upper. The minimum is capturing the loading readystate every time there is a page load. Nobody cares about the other ready states, because nothing important happens there. We replace the console and the global objects in the child window when we are in the loading ready state, and we add the unload listener there too...

The page error is connected to the later ready states too, so the minimalist approach should not contain that. 

```js
const engine = new ChromeEngine({
    opened: function (){},
    closed: function (){},
    request: function (){},
    responded: function (){},
    stop: function (){},
    stopped: function (){},
    error: function (){},
    unloaded: function (){},
    loading: function (){},
});
engine.open();
engine.isOpen();
engine.started();
engine.isStarted();
engine.stop();
engine.isStopped();
engine.close();
engine.isClosed();
engine.window;
engine.document;
engine.location;
```

It is not a big change but important. We have now 2-3 classes. 1 for navigation, 1 for the document and maybe 1 for the low level interface.

**Comment by inf3rno** [comment]
I am not sure the engine is a good name here. It could be ChromeNavigationEngine, but it could be ChromeWindow or ChromeNavigationManager, etc... too.

**Comment by inf3rno** [comment]
Finding the proper name is not a real issue, because it will be under the hood, so I can change the name of the class any time I want to.

**Comment by inf3rno** [comment]
In theory I can fix the wrong address error by sending an xhr to the same origin. If I get a response, then the address is ok and we got a CORS error, otherwise the address is not ok and we got a Wrong Address error. #126 This should be part of the upper ChromeWindowAdapter.

I managed to reduce the adapter interface to the following:

```
m = new WindowManager(id, h = hooks) -> h.opened(win)
// opening the window can be async

m.close() -> h.closed(win)
// closing the window is async too

m.crash(err) -> e2e.crashed(m, err) -> m[].crashed()
// should crash all windows by a single crash
// err: cant open, cant access, not loading, double listener call?

m.request(cb(win)=>evt) -> h.requested(evt)
// request should result an {address,type} event

m.requested(evt) -> h.requested(evt)
// requested can be called directly with a custom event if no stop needed

m.response(doc) -> h.responsed(doc) 
// the response contains the new prepared, verified doc in loading readystate

m.fail(err) -> h.failed(err) 
// navigation failed, cannot get the new doc, but the window should be reusable somehow
// err: unexpected, wrong address, already closed
```

Actually the latter is the hardest to solve, because if we cannot access Chrome error pages, then we won't be able to use location.assign() or any location function. I doubt that history is accessible, so we cannot do history.back() either, but we can check that later. All we can do is opening a new window, but we need to document, that history.back() wont give the expected results if we do so. There is no other way to fix this.

**Comment by inf3rno** [comment]
So the page error, unhandled rejection handling, replacing the logger, etc. should be all not adapter responsibility... This is cool this way. We can add an adapter test to make it sure, that there were no changes, so this is good for #133 too. If there was a change, that is affecting the lib, then we will be able to detect it almost instantly.

**Comment by inf3rno** [comment]
How would an `await navigator.load(address)` look like?

```js
p = Promise.exposed();
busy = true;
adapter.stop() ;
adapter.window.location.assign(address);
adapter.request();
this.hooks.responded = function (doc){
    // add console proxy
    // add error listener
    // add ready state change listener -> p.resolve(doc) by interactive
};
this.hooks.failed = function (err){
    // p.reject(err)
};
return p;
```

Hmm actually the hooks are ok if they are not redefined and using `this.p` instead of `p`.

So it will be just 

```js
busy = true;
adapter.stop() ;
adapter.window.location.assign(address);
adapter.request();
return this.p;
```

We can remove the stop() part by using the upper request/requested interface instead of a simplified one:

```js
busy = true;
adapter.request((win) => {
    win.location.assign(address);
});
return this.p;
```

actually we don't really need the request/requested by redirection:

```js
busy = true;
adapter.request();
return this.p;
```

this way the request will know, that the navigation is already started, and it just needs to follow it
the adapter does not have to know about the navigation type, etc... nobody cares about that on that low level...

okay, I think this will be the final interface.
so we can forget the adapter.stop() method
I don't want to deal with such details...

**Comment by inf3rno** [comment]
We could use events or callbacks instead of hooks, but I think the hooks are the ideal approach here, because they are created only once and they are connected to the actual promise through the context, so I can replace the promise any time... The event emitter is not ok here, because we have 1:1 relation between the adapter and the navigator, not 1:n. Giving a promise by the request would be ok too, so I could solve with a single method and a Promise instead of 3 hooks. The crash is interesting here. We could have 2 type of errors, 1 for crash 1 for navigation. So if it crashes, then we could throw that again by type. Window opening and closing is not something we care about. So I don't think we need those hooks here, just the methods.

```js
m = new e2e.WindowManager(id);
try {
    win = await m.request(cb); // lazy open
    prepare(win); // adding console, error handler, etc.
    p.resolve(win.document);
} catch(e){
    if (e instanceof Crash)
        throw e2e.crash(e);
    p.reject(e);
}
m.close();
```

**Comment by inf3rno** [comment]
I think it is perfect this way. I don't have to care about opening the window, it can be done by the first request. So it does not have to be sync. In theory we could even give the address to window.open, but there is no guarantee the macrotask finds the document in a loading ready state. We can try...

**Comment by inf3rno** [comment]
Now that we have a straightforward interface, we can write tests and implement it. We cannot tests for all of the environmental features, we need to trust the code more or less. Currently having a ChromeWindowManager is ok. Feature detection is not working here, because we cannot load data URI-s, so we cannot test for CORS errors without a real request. We could delay the testing to the first request, but for now having a Chrome solution only is ok for me. Actually I don't even want to name it as ChromeWindowManager, a simple WindowManager would be totally fine.

**Comment by inf3rno** [comment]
My only concern about the upper that it needs to await before it can return, so it might block the actual test. I need to benchmark this somehow to make sure that it works properly. I am not entirely sure how async functions work, but I suspect, that they return a promise before doing await. They are async after all and we call the synchronously. I guess.

**Comment by inf3rno** [comment]
```js
m = new e2e.WindowManager(id);
try {
    win = await m.request((win)=>win.location.assign(address)); // lazy open / auto stop
    prepare(win); // adding console, error handler, etc.
    p.resolve(win.document);
} catch(e){
    if (e instanceof Crash)
        throw e2e.crash(e);
    p.reject(e);
}
m.close();
```

I think the reduced test case is something like this:

```js
function a(){
    if (Math.random() > 0.5)
        return Promise.reject(new Error("r"));
    return Promise.resolve({q:1});
}

async function b (){
    p = Promise.exposed();
    try {
        x = await a();
        p.resolve(x);
    } catch(e) {
        p.reject(e);
    }
    return p;
};
```

We can benchmark it by adding a defer to a() and checking the result of b() by a sync call.

**Comment by inf3rno** [comment]
```js
    it("x", async function () {
        function a() {
            const r = createExposedPromise();
            setTimeout(r.resolve, 1000);
            return r;
        }

        async function b() {
            const p = createExposedPromise();
            try {
                const x = await a();
                p.resolve(x);
            } catch (e) {
                p.reject(e);
            }
            return p;
        }

        const timeOrigin = new Date().getTime();
        const dates = [];
        dates.push(new Date().getTime() - timeOrigin);
        const qp = a();
        const w = !!qp;
        dates.push(new Date().getTime() - timeOrigin);
        const q = await qp;
        dates.push(new Date().getTime() - timeOrigin);
        console.log(dates);
    });
```

```js
[0,0,1001]
```

**Comment by inf3rno** [comment]
Okay, so we can do the upper nested Promisified solution without the need of hooks.

**Comment by inf3rno** [comment]
In the navigator it would be `await adapter.unexpected()`. Otherwise it would be `adapter.unexpected((err) => {...})`. We should decide whether unexpected navigation is recoverable or not. I think it should be recoverable. So the app does a `window.stop()` and adds the unload handler to the new document. I am not sure whether it should call the other stuff on the document, probably that should be defined in the navigator whether it wants to run its own prepare or not.

**Comment by inf3rno** [comment]
The crash is not recoverable, because it means that the environment or the adapter code is not ok. It affects every adapter instance. The wrong address must be recoverable, but currently the only option to replace the window with a new one. That clears the history, so if it is combined with history tests, then it will result an unexpected failure.

**Comment by inf3rno** [comment]
Okay, the promise is not good for the adapter. We have recoverable and non-recoverable errors and not all of them can/should we send back through the promise. As far as I understand throwing an error in an async function means the same as `return Promise.reject(error)`.

**Comment by inf3rno** [comment]
We don't necessarily need new functions to set the callbacks. They are set only once, so we can do constructor injection.

```js
const adapter = new Adapter({
    success: function (window){...},
    failure: function (error){...}
});
adapter.start(function trigger(window){...});
adapter.close();
```

Well if we end up with this, then we have a problem. The success/failure will handle the async closing process too. So closing will go to success/failure too. Using promises might be better or maybe differently harder.

**Comment by inf3rno** [comment]
I thought about this, and we had a policy, that an unexpected navigation must cause an exception. A crash can cause an exception too, but it does not have to be something collective. So by a crash I don't have to notify the e2e lib, that the environment is not ok. If the dev calls the same method again and it crashes again, then that is ok. The notification would be some sort of recovery, but nowadays programmers don't care about that, because it is a lot of extra work without any real benefit.

So I think it is ok to send a rejection by crash instead of throwing some sort of uncatchable exception. It is ok to not handle unexpected navigation exception, but we need to somehow catch it for testing, and I don't want to add the unhandled rejection listeners etc. again. For now I won't care about that, but maybe later I'll add a hook to be able to log and ignore it.

```js
const adapter = new Adapter();
try {
    const page = await adapter.navigate(function trigger(window){...});
    await adapter.close(); // nem muszáj kivárni
} catch(e){}
```

Hmm for unexpected we could do something like this for testing:

```js
await adapter.unexpected();
```

Another option to set a hook for it:

```js
adapter.unexpected(function (){return shouldThrow;});
```

Yet another option to let the hook throw it:


```js
adapter.unexpected(function (err){ throw err; });
```

So if we replace it with an empty function, then it won't be thrown. If we want to log an throw it, then we can do that too. I like this one better than the others...

**Comment by inf3rno** [comment]
What about a general logger?

I don't think we need something like that in the adapter class. It is a higher level feature.

**Comment by inf3rno** [comment]
The adapter should not contain the ConsoleProxy either. The navigator can replace the console and everything else when it gets the loading document... All the adapter needs to do is making sure that we got a new document in a loading ready state before the scripts in its head run. This we can certainly do. We can call it loading adapter if we want to, it does not matter...

**Comment by inf3rno** [comment]
I  think this adapter interface design is worth another issue. It should be part of the low level API more or less, so we need to design it properly. Later we will use some sort of AdaptiveWindowAdapter, UnifiedWindowAdapter, MultiBrowserWindowAdapter, V8Adapter(engine based), etc...
